File: incre-tests/synduce/constraints/memo/max_sum_gt.f
Inductive Tree = leaf Int | node {Int, Tree, Tree};
Inductive TreeMemo = mleaf {Int, Int} | mnode {Int, Int, TreeMemo, TreeMemo};

memo = \t: TreeMemo.
  match t with
    mleaf {x, _} -> x
  | mnode {x, _, _, _} -> x
  end;

max = \a: Int. \b: Int. if < a b then b else a;

tmax = fix (
  \f: TreeMemo -> Int. \t: TreeMemo.
  match t with
    mleaf {_, x} -> x
  | mnode {_, x, l, r} -> max x (max (f l) (f r))
  end
);

is_memo = fix (
  \f: TreeMemo -> Bool. \t: TreeMemo.
  match t with
    mleaf {n, x} -> true
  | mnode {n, a, l, r} ->
    and (and (and (>= n (tmax l)) (>= n (tmax r))) (>= n a)) (and (f l) (f r))
  end
);

repr = fix (
  \f: TreeMemo -> Tree. \t: TreeMemo.
  match t with
    mleaf {n, a} -> leaf a
  | mnode {n, a, l, r} -> node {a, f l, f r}
  end
);

@Input key: Int;

spec = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf a -> if > a key then a else 0
  | node {a, l, r} -> + (if > a key then a else 0) (+ (f l) (f r))
  end
);

target = fix (
  \f: TreeMemo -> Compress TreeMemo. \t: TreeMemo.
  match t with
    mleaf {n, a} -> t
  | mnode {n, a, l, r} ->
    if > key n then t /*Avoid recursions*/
    else mnode {n, a, f l, f r}
  end
);

main =
  \mt: TreeMemo.
    if is_memo mt then spec (repr (target mt)) else 0;

File: ml-benchmark/synduce/constraints/memo/max_sum_gt.ml
@Input val key: int
type tree = Leaf of int | Node of int * tree * tree

type treememo = Mleaf of int * int | Mnode of int * int * treememo * treememo

let memo t =
  match t with
  | Mleaf (x, _) -> x
  | Mnode (x, _, _, _) -> x

let max a b = if a < b then b else a

let rec tmax t =
  match t with
  | Mleaf (_, x) -> x
  | Mnode (_, x, l, r) ->
      let ml = tmax l in
      let mr = tmax r in
      let m = max ml mr in
      max x m

let rec is_memo t =
  match t with
  | Mleaf (n, x) -> true
  | Mnode (n, a, l, r) ->
      let c1 = n >= tmax l in
      let c2 = n >= tmax r in
      let c3 = n >= a in
      let bl = is_memo l in
      let br = is_memo r in
      c1 && c2 && c3 && bl && br

let rec repr t =
  match t with
  | Mleaf (n, a) -> Leaf a
  | Mnode (n, a, l, r) ->
      let rl = repr l in
      let rr = repr r in
      Node (a, rl, rr)

let rec spec t =
  match t with
  | Leaf a ->
      if a > key then a else 0
  | Node (a, l, r) ->
      let v = if a > key then a else 0 in
      let sl = spec l in
      let sr = spec r in
      v + (sl + sr)

val target: treememo -> treememo compress
let rec target t =
  match t with
  | Mleaf (n, a) -> t
  | Mnode (n, a, l, r) ->
      if key > n then t
      else
        let tl = target l in
        let tr = target r in
        Mnode (n, a, tl, tr)

let program mt =
  if is_memo mt then
    let tm = target mt in
    let rt = repr tm in
    spec rt
  else 0